### PR TITLE
feature: Entity State polling

### PR DESCRIPTION
**Pull request by murrant**
Display entity state on the Inventory page.
Allows for alerting based on states.

image

DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

fixes: #7621

### CODE REVIEW

**Comment by laf** [comment]
So this isn't working for me, the sensors are added and then removed immediately.

However, is this not better to be added as yaml as it seems supported and then we can have a similar flag like rfc1628_compat in the os definition to load entity-sensor for OS we need it on?

**Comment by murrant** [comment]
Yeah, the valid was the wrong array or something.  This was just an exploration to see if we wanted this.

**Comment by laf** [comment]
I definitely say we should have it :)

**Comment by shimamizu** [comment]
screen shot 2017-11-13 at 4 34 10 pm
This looks promising :) Example on an Arista 7508 chassis.

**Comment by shimamizu** [comment]
Let me know what a simple alert rule would be to alert on say if PWR-3KT-AC-RED were to go down (aka if Power Supply in Slot 1 randomly died) based on where you're storing this in the database.

File: includes/polling/entity-state.inc.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                $state_data[$index]['entPhysical_id'] = $id;
+                $state_data[$index]['entStateLastChanged'] = $state['entStateLastChanged'];
+                d_echo("Inserting entity state:: ", '+');
+                d_echo($state);
+            }
+        }
+    }
+
+    if (!empty($state_data)) {
+        var_dump($state_data);
```
Comment (laf) [inline includes/polling/entity-state.inc.php]: Debug data left in.

**Comment by laf** [comment]
One big issue :(

image

Ignore the earlier data but the clean data from this PR is from 1pm onwards until it drops when I switch back from master. A change from 250 seconds to 50 seconds. I fear this will kill some devices :(

**Comment by murrant** [comment]
@laf yeah, any ideas on how to make that faster?  Like I said earlier, there is a lot of data here.
5 points of data for every piece of equipment in the entitymib (which includes ports usually)

**Comment by laf** [comment]
> @laf yeah, any ideas on how to make that faster? Like I said earlier, there is a lot of data here.
> 5 points of data for every piece of equipment in the entitymib (which includes ports usually)

Just do it how the current sensors table does it, multi gets until we've iterated through all the data. That means we need to either store the OIDs or be able to build them from the data we do store. If it's the same 5 OIDs with the index then you probably store that now.

**Comment by murrant** [comment]
I would think multi-getting the same data as the walk would be less efficient.

I guess there needs to be a way to disable the states for some entities?

**Comment by laf** [comment]
Some options:

  - walk entStateLastChanged, compare to the DB and for those that have changed, multi get the data you want.

  - do like we do for selected ports polling and walk entStateOper, if it's enabled multi get the data you want.

Also, is it worth storing the _prev values or are you using the StateLastChanged value to alert on or something else?

**Comment by laf** [comment]
For some reason this has stopped working for me. It's probably my local install so I'll keep digging.

**Comment by murrant** [comment]
Not sure about alerting.  I need to come up with an alert rule or two.  I'm not sure what a failed state looks like as I don't want to make my production test device fail ;)

**Comment by kkrumm1** [comment]
any word on this?

**Comment by murrant** [comment]
Ha sent gotten back to it.  Need to move existing code to discovery and write a more efficient poller.

**Comment by murrant** [comment]
@laf see if that is better for polling time

**Comment by laf** [comment]
It is, it still increases polling time from 35/40 seconds to 85 seconds.

I'd definitely be concerned about this for people though. That's enough to tip pollers over the 5 minute mark. What about setting this as disabled by default?

**Comment by murrant** [comment]
40s is a lot better that 200s.  Perhaps disabled by default is a good thing.

**Comment by laf** [comment]
> 40s is a lot better that 200s. Perhaps disabled by default is a good thing.

It is yes but it's (imho) too high for a default enable.

**Pull request merged**